ppcs/ppc0022-metaprogramming.md

```diff
@@ -0,0 +1,367 @@
+
+```perl
+$metasymbol = $metapackage->can_symbol($name);
+
+$metasymbol = $metapackage->get_symbol($name);
+```
+
+Returns a meta-symbol object instance to represent the symbol within the package, if such a symbol exists. If not either `undef` is returned or an exception is thrown.
+
+The exact type of symbol that is returned depends on the first character of the `$name` parameter - i.e. the sigil. Sigils of `*`, `$`, `@`, `%` and `%` respectively will return a meta-symbol representing a glob itself, or the scalar, array, hash or code slots from it.
+
+(Direct access to the scalar, array, hash or code slots of the glob is provided as a shortcut convenience for the common cases. Access to other more exotic slots of the glob such as filehandles or formats will have to be performed via the metaglob object itself).
+
+#### `add_symbol`
+
+```perl
+$metasymbol = $metapackage->add_symbol($name, $value);
+```
+
+Creates a new named symbol. If a value is passed it must be a reference to an item of the compatible type. If no initialisation is provided for variables, an empty variable is created. If no initialisation is provided for a subroutine then a forward declaration is created which has no body yet.
+
+#### `remove_symbol`
+
+```perl
+$metapackage->remove_symbol($name);
+```
+
+Removes a symbol. Nothing is returned.
+
+#### `list_symbols`
+
+```perl
+@metasymbols = $metapackage->list_symbols($filter)
+```
+
+Returns a list of symbols in the given package, in no particular order. (And in particular, we make no guarantees about the order of the results as compared to the order of the source code which declared it).
+
+TODO: Define what the filters look like. Need to be able to select scalar/array/hash/code as well as sub-stashes. Actually, do we need sub-stashes if we have `list_packages`?
+
+### Methods on Meta-Symbol Objects
+
+Several of the above methods return instances of a meta-symbol object. In practice, any meta-symbol object will represent either a glob (when using the "*" sigil), variable (when using the "$@%" sigils), or a subroutine (when using the "&"). Each of these cases is described further below. The following methods are available on all kinds.
+
+#### `basename`
+
+```perl
+$name = $metasymbol->basename;
+```
+
+Returns the final part of the symbol's fully qualified name; the name within the package after the final `::` but **excluding** the leading sigil.
+
+#### `sigil`
+
+```perl
+$sigil = $metasymbol->sigil;
+```
+
+Returns the single-character sigil that leads the symbol's name. This will be one of "*", "$", "@", "%" for variables or "&" for subroutines.
+
+#### `package`
+
+```perl
+$metapackage = $metasymbol->package;
+```
+
+Returns the meta-package object instance representing the package to which this symbol belongs.
+
+#### `name`
+
+```perl
+$name = $metasymbol->name;
+```
+
+Returns the fully-qualified name for this symbol, including its sigil and package name. This would be equivalent to
```

Comment:
Can you get all the names, here and above?

I could see that as being relatively expensive, but that's true for `for_reference()` too for SCALAR, ARRAY and HASH references

Reply:
It's not too bad. You just walk the `->parent` chain collecting up the names. I'm fairly sure core perl does that in places. Plus it doesn't matter if it's a bit expensive; you only need to do it in the accessor itself if actually requested. I could imagine most use-cases wouldn't need it.

Reply:
I'm confused, I don't see a simple way to get from:
```
my $ref = \$My::Deep::Package::x; # 
my $metasymbol = meta::for_reference($ref);
```
to the position in the symbol table.

The scalar / array / hash doesn't include any name information that could be used to find the stash, so there's nothing to follow parent links on.

Perhaps objects from for_reference() don't support returning the name, which might be where I was confused.

Reply:
> I'm confused, I don't see a simple way to get from:
...

Oh, hrm. I thought perhaps that package variables all reference their GV, but maybe they don't? Perhaps it's only CVs that allow an easy backward path back to the name + package.

If not, then I guess unless we add that then indeed doing a `for_reference` would have to just give you a meta-symbol around an "anonymous" variable. It might well have a name but we don't happen to know it.